openmc/mgxs/mgxs.py

```diff
@@ -0,0 +1,5909 @@
+        for compressed data storage
+    loaded_sp : bool
+        Whether or not a statepoint file has been loaded with tally data
+    derived : bool
+        Whether or not the MGXS is merged from one or more other MGXS
+    hdf5_key : str
+        The key used to index multi-group cross sections in an HDF5 data store
+
+    """
+
+    def __init__(self, domain=None, domain_type=None,
+                 energy_groups=None, by_nuclide=False, name='', num_polar=1,
+                 num_azimuthal=1):
+        self._name = ''
+        self._rxn_type = None
+        self._by_nuclide = None
+        self._nuclides = None
+        self._estimator = 'tracklength'
+        self._domain = None
+        self._domain_type = None
+        self._energy_groups = None
+        self._num_polar = 1
+        self._num_azimuthal = 1
+        self._tally_trigger = None
+        self._tallies = None
+        self._rxn_rate_tally = None
+        self._xs_tally = None
+        self._sparse = False
+        self._loaded_sp = False
+        self._derived = False
+        self._hdf5_key = None
+        self._valid_estimators = ESTIMATOR_TYPES
+
+        self.name = name
+        self.by_nuclide = by_nuclide
+
+        if domain_type is not None:
+            self.domain_type = domain_type
+        if domain is not None:
+            self.domain = domain
+        if energy_groups is not None:
+            self.energy_groups = energy_groups
+        self.num_polar = num_polar
+        self.num_azimuthal = num_azimuthal
+
+    def __deepcopy__(self, memo):
+        existing = memo.get(id(self))
+
+        # If this is the first time we have tried to copy this object, copy it
+        if existing is None:
+            clone = type(self).__new__(type(self))
+            clone._name = self.name
+            clone._rxn_type = self.rxn_type
+            clone._by_nuclide = self.by_nuclide
+            clone._nuclides = copy.deepcopy(self._nuclides, memo)
+            clone._domain = self.domain
+            clone._domain_type = self.domain_type
+            clone._energy_groups = copy.deepcopy(self.energy_groups, memo)
+            clone._num_polar = self._num_polar
+            clone._num_azimuthal = self._num_azimuthal
+            clone._tally_trigger = copy.deepcopy(self.tally_trigger, memo)
+            clone._rxn_rate_tally = copy.deepcopy(self._rxn_rate_tally, memo)
+            clone._xs_tally = copy.deepcopy(self._xs_tally, memo)
+            clone._sparse = self.sparse
+            clone._loaded_sp = self._loaded_sp
+            clone._derived = self.derived
+            clone._hdf5_key = self._hdf5_key
+
+            clone._tallies = OrderedDict()
+            for tally_type, tally in self.tallies.items():
+                clone.tallies[tally_type] = copy.deepcopy(tally, memo)
+
+            memo[id(self)] = clone
+
+            return clone
+
+        # If this object has been copied before, return the first copy made
+        else:
```

Comment:
Definitely a style decision but it would probably be easier to read if you immediately returned `existing` when you found it in the memoization cache.